planning/trajectory_optimization/gcs_trajectory_optimization.h

```diff
@@ -0,0 +1,284 @@
+    trajectories::BezierCurve<symbolic::Expression> u_r_trajectory_;
+
+    friend class GCSTrajectoryOptimization;
+  };
+
+  class EdgesBetweenSubgraphs final {
+   public:
+    DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(EdgesBetweenSubgraphs);
+
+    ~EdgesBetweenSubgraphs() = default;
+
+   private:
+    EdgesBetweenSubgraphs(const Subgraph* from, const Subgraph* to,
+                          const geometry::optimization::ConvexSet* subspace,
+                          GCSTrajectoryOptimization* gcs);
+
+    bool RegionsConnectThroughSubspace(
+        const geometry::optimization::ConvexSet& A,
+        const geometry::optimization::ConvexSet& B,
+        const geometry::optimization::ConvexSet& subspace);
+
+    GCSTrajectoryOptimization* gcs_;
+
+    std::vector<geometry::optimization::GraphOfConvexSets::Edge*> edges_;
+
+    // We keep track of the edge variables and trajectory since other
+    // constraints and costs will use these.
+    VectorX<symbolic::Variable> u_duration_;
+    VectorX<symbolic::Variable> u_vars_;
+    trajectories::BezierCurve<symbolic::Expression> u_r_trajectory_;
+
+    VectorX<symbolic::Variable> v_duration_;
+    VectorX<symbolic::Variable> v_vars_;
+    trajectories::BezierCurve<symbolic::Expression> v_r_trajectory_;
+
+    friend class GCSTrajectoryOptimization;
+  };
+
+  /** Returns the number of position variables. */
+  int num_positions() const { return num_positions_; }
+
+  /**
+  @param show_slacks determines whether the values of the intermediate
+  (slack) variables are also displayed in the graph.
+  @param precision sets the floating point precision (how many digits are
+  generated) of the annotations.
+  @param scientific sets the floating point formatting to scientific (if true)
+  or fixed (if false).
+  */
+  std::string GetGraphvizString(bool show_slack = true, int precision = 3,
+                                bool scientific = false) const {
+    return gcs_.GetGraphvizString(std::nullopt, show_slack, precision,
+                                  scientific);
+  }
+
+  /** Creates a Subgraph with the given regions and indices.
+  @param regions represent the valid set a control point can be in.
+    We retain copy of the regions since other functions may access them.
+  @param edges_between_regions is a list of pairs of indices into the regions
+    vector. For each pair representing an edge between two regions, an edge is
+  added within the subgraph. Note that the edges are directed so (i,j) will only
+  add an edge from region i to region j.
+  @param order is the order of the Bézier curve.
+  @param d_max is the maximum duration to spend in a region (seconds). Some
+  solvers struggle numerically with large values.
+  @param d_min is the minimum duration to spend in a region (seconds) if that
+  region is visited on the optimal path.
+    Some cost and constraints are only convex for d > 0. For example the
+  perspective quadratic cost of the path energy ||ṙ(s)||² / d becomes non-convex
```

Comment:
This is a bit imprecise (one should take the limit and also see what happens to r_dot as d goes to zero). I would just say that we don't want regions that are visited for zero seconds, as it never makes sense? Also using `d` for the durations of a trajectory might be confusing. Typically one uses T or h. While d is reserved to the space dimension or the degree.